service/src/main/java/greencity/service/utility/EntityManagerUtils.java

```diff
@@ -0,0 +1,413 @@
+
+    static {
+        StringBuilder builder = new StringBuilder();
+        builder.append("\\s*");
+        builder.append("(select\\s+((distinct)?((?s).+?)?)\\s+)?(from\\s+");
+        builder.append("[._$[\\P{Z}&&\\P{Cc}&&\\P{Cf}&&\\P{Punct}]]+");
+        builder.append("(?:\\s+as)?\\s+)");
+        builder.append(IDENTIFIER_GROUP);
+        builder.append("(.*)");
+        COUNT_MATCH = Pattern.compile(builder.toString(), Pattern.DOTALL | Pattern.CASE_INSENSITIVE);
+        builder = new StringBuilder();
+        builder.append("(?<=\\bfrom)");
+        builder.append("(?:\\s)+");
+        builder.append(IDENTIFIER_GROUP);
+        builder.append("(?:\\sas)*");
+        builder.append("(?:\\s)+");
+        builder.append("(?!(?:where|group\\s*by|order\\s*by))(\\w+)");
+        ALIAS_MATCH = Pattern.compile(builder.toString(), Pattern.CASE_INSENSITIVE);
+    }
+
+    public static final String ENTITY_GRAPH_ARGUMENT_EXCEPTION = "One or more specified attributes can't be applied";
+
+    private record Key(String queryString, String countProjection, boolean nativeQuery) {
+    }
+
+    private final LoadingCache<Key, String> countQueryStringCache;
+
+    public EntityManagerUtils(@Value("${greencity.cache.lifetime}") long cacheLifeDuration) {
+        countQueryStringCache = Caffeine.newBuilder()
+            .maximumSize(100)
+            .expireAfterWrite(cacheLifeDuration, TimeUnit.MINUTES)
+            .build(
+                key -> createCountQueryStringForInternal(key.queryString(), key.countProjection(), key.nativeQuery()));
+    }
+
+    /**
+     * Methods creates EntityGraph which can be then set as fetchgraph or loadgraph
+     * query hint. This method is well-suited and proves effectiveness in fixing n+1
+     * via eager loading of chosen manyToOne or oneToOne relations, while only slows
+     * down a query in case of collection-type relations usage due to cartesian
+     * product emerging.
+     *
+     * @param entityClass class of Entity that entityGraph is built for;
+     * @param attributes  names of relation fields to be fetched, order is important
+     *                    in case of nested relations, you can't specify a.b.c
+     *                    before specifying a.b;
+     * @return {@link EntityGraph} ready-to-use EntityGraph object with all
+     *         attributes and subgraphs set.
+     * @author Oleksandr Ilnytskyi
+     */
+    public <T> EntityGraph<T> createEntityGraph(Class<T> entityClass, List<String> attributes) {
+        EntityGraph<T> entityGraph = entityManager.createEntityGraph(entityClass);
+
+        Map<String, Subgraph<?>> subgraphsMap = new HashMap<>();
+
+        for (String attribute : attributes) {
+            if (!attribute.contains(".")) {
+                try {
+                    entityGraph.addAttributeNodes(attribute);
+                } catch (Exception exception) {
+                    throw new IllegalStateException(ENTITY_GRAPH_ARGUMENT_EXCEPTION);
```

Comment:
Please preserve the original exception message and cause when rethrowing. Currently, the root cause is lost, which makes debugging harder.